### DISCUSSION TITLE
Getting a single output while running in parallel

### DISCUSSION THREAD

**Discussion by dheligman** [NanoComp/meep#1999] [Q&A] [answered]
Hi All,

I have been using meep mpi and am running into an issue where the python could is outputting a plot, or saving a file for each node I am using for my simulation. Is there a function that allows for only one copy of the output to be saved?

**Comment by oskooi** [comment, marked as answer]
> Is there a function that allows for only one copy of the output to be saved?

In a parallel Meep simulation consisting of multiple chunks, the output functions (e.g., `output_epsilon`, `output_efield`, etc.) automatically write a *single* HDF5 file to disk. 

If you are trying to write your own data (eg., `numpy.savez`, etc) or save a plot to disk (e.g., `matplotlib.pyplot.savefig`, etc) as part of a parallel simulation, each MPI process will redundantly write its own copy to disk unless you preface the write statement with `if mp.am_master()` as described in https://meep.readthedocs.io/en/latest/Parallel_Meep/#technical-details:

```
Sometimes you only want an operation to take place on one process. A common use case is showing a matplotlib
plot with plt.show(), or saving a file with plt.savefig(). In cases where you need to distinguish different MPI 
processes in your Python/Scheme file, you can use the following functions:

meep.am_master(), (meep-am-master) — Returns true if the current process is the master process (rank 0).

This can be useful for calling external I/O or visualization routines, e.g. Matplotlib plotting functions, that you 
only want to execute on the master process. Note that the Scheme (print) or Python print functions are already
set up so that by default their output is suppressed on non-master processes.
```

**Comment by Niccolo-Marcucci** [comment]
The solution I found for your particular problem is to test if the figure contains an image, before saving it.
```
    fig = plt.figure(dpi=100)
    plot = sim.plot2D()
    try:
        fig.colorbar(plot.images[0])
    except:
        plt.close()
        print("Only one of the parallel jobs jobs will print the image")
    else:
        fig.savefig(f'section-xy.jpg')
        plt.close()
```
`plot.images[0]`  raises an error if the figure is empty (as it is in all non master jobs in a parallels simulation).

I've never tried, but I guess you can simply use the function `meep.am_master()` and create the figure only if it returns `True`
